### PR TITLE
Add FieldDataContext

### PR DESCRIPTION
**Pull request by romseygeek**
MappedFieldType#fieldDataBuilder() currently takes two parameters, a fully qualified
index name and a supplier for a SearchLookup.  We expect to add more parameters here
as we add support for loading fielddata from source.  Rather than telescoping the
parameter list, this commit instead introduces a new FieldDataContext carrier object
which will allow us to add to these context parameters more easily.

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/index/mapper/IndexFieldMapper.java [hunk 1]
```diff
@@ -88,6 +87,11 @@ public List<Object> fetchValues(SourceLookup lookup, List<Object> ignoredValues)
                 }
             };
         }
+
+        @Override
+        public boolean eagerGlobalOrdinals() {
+            return false;
```
Comment (nik9000) [inline server/src/main/java/org/elasticsearch/index/mapper/IndexFieldMapper.java:93]: Is this related?

File: test/framework/src/main/java/org/elasticsearch/index/mapper/MapperServiceTestCase.java [hunk 2]
```diff
@@ -339,6 +340,8 @@ protected final XContentBuilder runtimeFieldMapping(CheckedConsumer<XContentBuil
         });
     }
 
+    protected static FieldDataContext EMPTY_CONTEXT = new FieldDataContext("test", null);
```
Comment (nik9000) [inline test/framework/src/main/java/org/elasticsearch/index/mapper/MapperServiceTestCase.java:343]: Is it better to use `staticContext` here?

I wonder about the name `static`, btw. The name gets me mixed up with the `static` keyword.

**Review by nik9000 — COMMENTED** [review]
I very much like it.

**Review by jdconrad — APPROVED** [review]
@romseygeek Thank you for doing this refactoring! It looks good to me, but I would be more comfortable if @nik9000 or someone else more familiar with this area of code approved as well prior to committing.

File: test/framework/src/main/java/org/elasticsearch/index/mapper/MapperServiceTestCase.java (hunk 2 above)
Comment (jdconrad) [reply test/framework/src/main/java/org/elasticsearch/index/mapper/MapperServiceTestCase.java:343]: Maybe `empty` is a better name than `static` for the method as well?

**Comment by nik9000** [comment]
I approve.

File: server/src/main/java/org/elasticsearch/index/mapper/IndexFieldMapper.java (hunk 1 above)
Comment (romseygeek) [reply server/src/main/java/org/elasticsearch/index/mapper/IndexFieldMapper.java:93]: The parent value is false as well, so it's not really necessary, but it clarifies that this won't be called during index warming and so we don't need the fully qualified index name and can happily use an empty fielddatacontext at that point.

File: test/framework/src/main/java/org/elasticsearch/index/mapper/MapperServiceTestCase.java (hunk 2 above)
Comment (romseygeek) [reply test/framework/src/main/java/org/elasticsearch/index/mapper/MapperServiceTestCase.java:343]: I've renamed to `noScriptContext`, as the primary usecase seems to be places where we don't support runtime scripts in fielddata.

File: server/src/main/java/org/elasticsearch/index/mapper/IndexFieldMapper.java (hunk 1 above)
Comment (nik9000) [reply server/src/main/java/org/elasticsearch/index/mapper/IndexFieldMapper.java:93]: Maybe a comment and make it `final`?

**Review by nik9000 — APPROVED** [review]

File: server/src/main/java/org/elasticsearch/index/mapper/IndexFieldMapper.java (hunk 1 above)
Comment (romseygeek) [reply server/src/main/java/org/elasticsearch/index/mapper/IndexFieldMapper.java:93]: It's on a final class so checkstyle whinges if the method is marked as final 🤷

File: modules/mapper-extras/src/main/java/org/elasticsearch/index/mapper/extras/MatchOnlyTextFieldMapper.java [hunk 3]
```diff
@@ -276,7 +275,7 @@ public Query phrasePrefixQuery(TokenStream stream, int slop, int maxExpansions,
         }
 
         @Override
-        public IndexFieldData.Builder fielddataBuilder(String fullyQualifiedIndexName, Supplier<SearchLookup> searchLookup) {
+        public IndexFieldData.Builder fielddataBuilder(FieldDataContext fieldDataContext) {
```
Comment (javanna) [inline modules/mapper-extras/src/main/java/org/elasticsearch/index/mapper/extras/MatchOnlyTextFieldMapper.java:278]: bikeshedding: fieldData of fielddata?

File: server/src/main/java/org/elasticsearch/index/IndexService.java [hunk 4]
```diff
@@ -205,7 +206,7 @@ public IndexService(
                 this.indexSortSupplier = () -> indexSettings.getIndexSortConfig()
                     .buildIndexSort(
                         mapperService::fieldType,
-                        (fieldType, searchLookup) -> indexFieldData.getForField(fieldType, indexFieldData.index().getName(), searchLookup)
+                        (fieldType, searchLookup) -> indexFieldData.getForField(fieldType, FieldDataContext.noRuntimeFields("index sort"))
```
Comment (javanna) [inline server/src/main/java/org/elasticsearch/index/IndexService.java:209]: IndexSortConfig throws an exception too by providing a supplier of search lookup that throws exception. Which exception comes first and do we need both?

**Comment by javanna** [comment]
One question I had: could you add more context on what the plan we have in mind for loading fielddata from source? What additional parameters do we plan to add for instance that make this refactoring needed? Thanks!

**Comment by nik9000** [comment]
Ah! Well then. A comment would be lovely. But otherwise it's cool.
